### DISCUSSION TITLE
crossfade sample code

### DISCUSSION THREAD

**Discussion by Nicolas-Imago** [savonet/liquidsoap#3838] [User Support]
Hi guys, for a web radio project we implemented a very simple radio with liquidsoap : 

   playlist = playlist("/var/www/html/mp3")
   live = input.http("http://localhost:8000/live.mp3") 
   radio = fallback(track_sensitive=false, [live, playlist])
   output.icecast(%mp3, host = "localhost", port = 8000, password = "PWD", mount = "webradio.mp3", mksafe(radio))

The idea is to have a playlist (all day long) that we can interrupt at any time by a live. 
And it works perfectly ! 

But now we would like to improve adding a fading transition from playlist to live and from live to playlist. 

We tried to find help here and had a look to the documentation. 
But we could not found. 

We tried (advised by Chat GPT) to add : 
   radio_fading = crossfade(start_next=true, fade_in=2.0, fade_out=2.0, [live, playlist])
and changed last line as : 
   output.icecast(%mp3, host = "localhost", port = 8000, password = "PWD", mount = "webradio.mp3", mksafe(radio_fading))

but it doesn't work.
I guess solution is very simple ?

Nicolas

**Comment by Moonbase59** [comment]
Never trust an "AI" to generate working code! There is no `start_next` in `crossfade`. Docs here → https://www.liquidsoap.info/doc-dev/reference.html#crossfade, _Cookbook_ here → https://www.liquidsoap.info/doc-dev/cookbook.html.

Then there’s also _"The Book"_—really worth reading. → https://www.liquidsoap.info/doc-dev/book.html

**Comment by Nicolas-Imago** [comment]
Thanks Moonbase59 for your help. 

You right, with Chat GPT implementation, I had an error about "start_next", so I removed it yesterday but I had another error message 

_At mux.liq, line 38, char 55-71:
   radio_fading = crossfade(fade_in=2.0, fade_out=2.0, [live, playlist])

Error 5: this value has type
  [_]
but it should be a subtype of the type of the value at /usr/share/liquidsoap/2.1.3/libs/fades.liq, line 369, char 58
  source(_)_

I also read the 2 links you suggest (before to post my message yesterday) and I could not understand how to implement. 
Unfortunately, there is no sample :(

Do you know the line I should implement to replace : 
   radio_fading = crossfade(fade_in=2.0, fade_out=2.0, [live, playlist])

Thanks a lot in advance
Nicolas

**Comment by Nicolas-Imago** [comment]
Hi ! Moonbase59, any other helper to help us ? ;)

**Comment by Moonbase59** [comment]
Well, you already have a `fallback` in your original code, which supports specifying transition functions, so you could easily add one each for the transitions _playlist → live_ and _live → playlist_, [as shown in the Cookbok](https://www.liquidsoap.info/doc-dev/cookbook.html#play-a-short-silence-when-transitioning-out-of-input.harbor):

```ruby
# Transition to live
def to_live(playlist, live) =
  #sequence([playlist,live])
  add([fade.in(duration=2.0, live), fade.out(duration=2.0, playlist)])
end

# Transition back to playlist
def to_playlist(live, playlist) =
  #sequence([live, playlist])
  add([fade.in(duration=2.0, playlist), fade.out(duration=2.0, live)])
end

radio = fallback(
  track_sensitive=false,
  transitions=[to_live, to_playlist],
  [live, playlist]
)
```

Out of my head, not tested. I just replaced the suggested `sequence` from the Cookbook with an [`add`](https://www.liquidsoap.info/doc-dev/reference.html#add) that allows the fades to be specified.

— or —

If you really wanted to use a [`crossfade`](https://www.liquidsoap.info/doc-dev/reference.html#crossfade) instead, which would _also_ crossfade the playlist tracks, just use the source instead of a list of sources (your `Error 5: []`):

```ruby
radio = crossfade(fade_in=2.0, fade_out=2.0, radio)
```

Notice the _subtle differences_:

- `fade.in` in the `add` is a _function_, while
- `fade_in` in the `crossfade` is just a _parameter_ given to the `crossfade` function.

Learn Liquidsoap, it’s not _so_ easy but eventually pays, and you can do _anything_!

**Reply by Nicolas-Imago** [reply]
Thanks for your help. 
I couldn't make work crossfade with your suggested code, I still have the same error 5. 

But I managed to test your suggested code with "add" function. I works, but it seems not to work as it should. 

I also found the book here : https://www.liquidsoap.info/book/book.pdf 
I started to read it to understand better how crossfade works.

Thanks.
